Complete pluggable trimesh_curvature in compas_rhino

XingxinHE opened the pull request:
Hi compas team, 

I am submitting my PR.

### What type of change is this?

- [ ] Bug fix in a **backwards-compatible** manner.
- [ x] New feature in a **backwards-compatible** manner.
- [ ] Breaking change: bug fix or new feature that involve incompatible API changes.
- [ ] Other (e.g. doc update, configuration, etc)

### Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x ] I added a line to the `CHANGELOG.md` file in the `Unreleased` section under the most fitting heading (e.g. `Added`, `Changed`, `Removed`).
- [ x] I ran all tests on my computer and it's all green (i.e. `invoke test`).
- [ ] I ran lint on my computer and there are no errors (i.e. `invoke lint`).
- [x ] I added new functions/classes and made them available on a second-level import, e.g. `compas.datastructures.Mesh`.
- [ x] I have added tests that prove my fix is effective or that my feature works.
- [x ] I have added necessary documentation (if appropriate)


There are 3 things missing in my PR.

One is `invoke lint`, although I use `linter` in vscode to debug... somehow there are still error pops up like:

1#
`undefined name 'pluggable'`, I reference the script in `src/compas_rhino/geometry/booleans/__init__.py` while I don't know how to skip this error.
2#
`invalid escape sequence '\s'`, I reference `src/compas/datastructures/mesh/curvature.py` to write down some formula. I suppose that is similar to LaTex. However, I cannot pass the lint with all the formula using `.. math::` or `:math:`.

3#
`invoke docs`. For this part, I used to manage to invoke docs. But this time, it said:
```
Running Sphinx v2.4.3

Configuration error:
There is a programmable error in your configuration file:

Traceback (most recent call last):
  File "c:\users\xingx\anaconda3\envs\compas-dev\lib\site-packages\sphinx\config.py", line 348, in eval_config_file
    execfile_(filename, namespace)
  File "c:\users\xingx\anaconda3\envs\compas-dev\lib\site-packages\sphinx\util\pycompat.py", line 81, in execfile_
    exec(code, _globals)
  File "C:\E-Architecture\Computational Design\compas\docs\conf.py", line 12, in <module>
    import m2r2
ModuleNotFoundError: No module named 'm2r2'
```



OK. Now I describe my contribution. I worked on the plugin of `trimesh_curvature` in `compas_rhino` using solely RhinoCommon SDK which are:
`trimesh_gaussian_curvature`, already defined in `src/compas/geometry/trimesh/curvature.py`
`trimesh_mean_curvature`, not defined yet, but required to compute principal curvature
`trimesh_principal_curvature`, already defined in `src/compas/geometry/trimesh/curvature.py`

In light of these curvatures, I may have few questions.

a# Do you have any plans on `vector curvature`? If so, can I put them into the current file?
Currently, in `src/compas/geometry/trimesh/curvature.py`, the Gaussian is scalar curvature while the principal curvature is marked with `direction` a.k.a vector curvature. I see there is a mix here. Can I write both `vector curvature` and `scalar curvature` here?

b# Where should the barycentric area function goes?
When computing curvature, I may need `vertex barycentric area` function. However, this is not the only option for discrete setting. Do you have a plan on computing vertex area? Like `vertex voronoi area` and `mixed voronoi area`. Can I put something like this ?
`src/compas/geometry/trimesh/area.py`
Or do you have a better suggestion?

That's all!

I wish you all a nice weekend.

Xingxin

XingxinHE committed:
initialize a function of mesh_to_pointscloud

XingxinHE committed:
Revert "initialize a function of mesh_to_pointscloud"

This reverts commit 5cdb4e119d0b1453216a4da357265d54cbef5f7b.

XingxinHE committed:
Merge remote-tracking branch 'upstream/main'

XingxinHE committed:
Merge remote-tracking branch 'upstream/main'

XingxinHE committed:
Complete Feature trimesh_gaussian_curvature in Rhino

The pluggable function trimesh_gaussian_curvature is completed in Rhino
environment using purely RhinoCommon SDK.

Related formula, description, and example are also added to the file.
The function has been tested using GH_Python in Grasshopper environment.

XingxinHE committed:
Minor Fixed: correct 'teta' to 'theta' in current doc

XingxinHE committed:
Modified trimesh_gaussian_curvature

1. Fix some lint error.
2. Detect if the input is already Rhino.Geometry.Mesh
3. Clearer logic on the vertices and topology_vertices

XingxinHE committed:
Finished feature, trimesh_mean_curvature in compas_rhino

The discrete mean curvature of such vertex is 1/4 of the sum connected
edges and their dihedral angle.

XingxinHE committed:
Finished feature trimesh_principle_curvature in compas_rhino

The principle curvature currently is scalar curvature. The dual cell
area is computed by barycentric area which is with less computing time.

Future implementation could be vector curvature of principle curvature.

XingxinHE committed:
Add spot.obj to data directory for it is an opensource mesh and it is useful for testing trimesh function

XingxinHE committed:
Update CHANGLOG

XingxinHE committed:
Fix version in CHANGELOG

XingxinHE committed:
Minor correction, correct principle to principal

XingxinHE committed:
Fix mean curvature with correct dihedral angle
1. The last version is over simplified which does not consider the face
   orientation. It leads to a non-negative curvature that is wrong.
2. Now the dihedral angle has fully taken the face orientation into
   consideration.[right hand rule]

XingxinHE commented:
Hi COMPAS team, I hope you all have fun in Venezia! I modify a bit of the algorithm to get the correct dihedral angle. I enclose 2 screenshots for you better understand the curvature.

![bunny](https://user-images.githubusercontent.com/30434598/119507342-56e27c80-bda1-11eb-992a-4784f1ae1534.png)
![Spot](https://user-images.githubusercontent.com/30434598/119507353-59dd6d00-bda1-11eb-88a4-e296201728fa.png)

tomvanmele commented:
hi @XingxinHE 

looking good so far!
we will try to have a look as soon as possible...

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,316 @@
+    Formula:
+    .. math::
+
+        K_{i} = 2\pi-\sum\theta_{i}^{jk}
```
Do you have references for this and the other formulas?  It would be nice to include in the docs links to explanations of the theory behind the methods.

XingxinHE replied:
Hi Dr. Lytle,

thanks for the suggestion. I will include something like this under all formulas:

```
    References
    ----------
    .. [1] Discrete Curvature by Keenan Crane, 03:16-07:11, at https://youtu.be/sokeN5VxBB8
```

Sounds good?

XingxinHE committed:
Added references to formulas

There are 3 formulas included in this script that are discrete gaussian, mean, and principal curvature. References have been added accompanying with detailed explanation.

beverlylytle commented on CHANGELOG.md:
```diff
@@ -10,18 +10,61 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
+## Unreleased
+
+### Added
+
```
```suggestion
```
These lines shouldn't be duplicated.

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Description: The angle defect at a vertex is used to describe the Gaussian curvature in a neighborhood around a vertex.
+
+    Notation Convention:
+    :math:`K` - discrete Gaussian curvature
```
```suggestion
    :math:`K_{i}` - discrete Gaussian curvature at vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    """
+    # (1) see if input is already Rhino.Geometry.Mesh
+    mesh = Rhino.Geometry.Mesh()
+    if type(M) != Rhino.Geometry.Mesh:
```
```suggestion
    if not isinstance(M, Rhino.Geometry.Mesh):
```
Sometimes `type` can give unexpected results...

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+        mesh = M
+
+    # (2) Prepare ingredient and return list
+    pi_2 = 2 * pi
```
There's a movement to call this value tau : https://en.wikipedia.org/wiki/Turn_(angle)#Tau_proposals, but `pi_2` is totally reasonable name for this variable.

XingxinHE replied:
This is fun. Thanks for the science popularization from a math perspective! Good to know : )

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+                mesh.Faces.AddFace(*face)
+    else:
+        mesh = M
+
```
Here I feel there should be a check on whether `M` is really a trimesh (and similarly for the other pluggables).  The method will return a result even if `M` is not a trimesh, but it's not a sensical result.   Or more exactly, I believe discrete gaussian curvature does make sense for any polygonal mesh, but I don't think mean (vertex) curvature does nor does the barycentric area (and hence principal curvature also does not).

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Description: The discrete mean curvature is computed by edge length and its dihedral angle.
+
+    Notation Convention:
+    :math:`H` - discrete mean curvature
```
```suggestion
    :math:`H_{i}` - discrete mean curvature at vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    return areas
+
+
+def dihedral_angle(e, n1, n2):
```
Maybe this could also use a reference? perhaps https://youtu.be/NlU1m-OfumE

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+
+    References
+    ----------
+    .. [1] Formula of Discrete Mean Curvature available at Keenan Crane's lecture, 03:16-07:11, at https://youtu.be/sokeN5VxBB8
```
In the video, this is actually the "vertex mean curvature" (where as the "mean curvature" is associated to an edge instead of a vertex).

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Description: The discrete principal curvature is computed by mean curvature, Gaussian curvature, and vertex area.
+
+    Notation Convention:
+    :math:`\kappa_1, \kappa_2` - The max principal curvature and the min principal curvature
```
```suggestion
    :math:`\kappa^1_i, \kappa^2_i` - The max principal curvature and the min principal curvature at the vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Formula:
+    .. math::
+
+        \frac{H_i}{A_i}\pm\sqrt{( \,\frac{H_i}{A_i})\,^2-\frac{K_i}{A_i}}
```
```suggestion
       \kappa^1_i, \kappa^2_i =  \frac{H_i}{A_i}\pm\sqrt{( \,\frac{H_i}{A_i})\,^2-\frac{K_i}{A_i}}
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+        K_ = K[i] / A[i]
+
+        k1.append(H_ + sqrt(max(0, H_ * H_ - K_)))
+        k2.append(H_ - sqrt(max(0, H_ * H_ - K_)))
```
`sqrt` is expensive, and to do this twice for every vertex in a mesh might be slower than necessary.  
```suggestion
        discriminant = sqrt(max(0, H_ * H_ - K_))
        k1.append(H_ + discriminant)
        k2.append(H_ - discriminant)
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+        vert_neighbors_topo = mesh.TopologyVertices.ConnectedTopologyVertices(mesh.TopologyVertices.TopologyVertexIndex(i), True)
+        vert_neighbors = []
+        if vert_neighbors_topo is None:
+            K.append(0)
```
I'm not sure I agree with zero being the default value here.  Saying that it's zero would suggest that the mesh is locally flat, when really I think the curvature is undefined (maybe `None`?) at such locations.  This comment applies to all the pluggables.

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+
+    Notation Convention:
+    :math:`H` - discrete mean curvature
+    :math:`E` - all the edges connect vertex i
```
```suggestion
    :math:`E` - all the edges connected to vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Notation Convention:
+    :math:`H` - discrete mean curvature
+    :math:`E` - all the edges connect vertex i
+    :math:`j` - the vertex connects vertex i
```
```suggestion
    :math:`j` - a vertex connected to vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Notation Convention:
+    :math:`K` - discrete Gaussian curvature
+    :math:`j,k` - the vertices from the Star of vertex i
+    :math:`e_{ij}, e_{ik}` - the vector from vertex i to j, i to k
```
```suggestion
    :math:`e_{ij}, e_{ik}` - the vectors from vertex i to j and i to k
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+
+    Notation Convention:
+    :math:`\kappa_1, \kappa_2` - The max principal curvature and the min principal curvature
+    :math:`H_i` - the discrete mean curvature of vertex i
```
```suggestion
    :math:`H_i` - the discrete mean curvature at vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    Notation Convention:
+    :math:`\kappa_1, \kappa_2` - The max principal curvature and the min principal curvature
+    :math:`H_i` - the discrete mean curvature of vertex i
+    :math:`K_i` - the discrete Gaussian curvature of vertex i
```
```suggestion
    :math:`K_i` - the discrete Gaussian curvature at vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    :math:`\kappa_1, \kappa_2` - The max principal curvature and the min principal curvature
+    :math:`H_i` - the discrete mean curvature of vertex i
+    :math:`K_i` - the discrete Gaussian curvature of vertex i
+    :math:`A_i` - the vertex area of vertex i
```
```suggestion
    :math:`A_i` - the area of the dual cell centered at vertex i
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+]
+
+
+@pluggable(category="trimesh", requires=['Rhino'])
```
```suggestion
@plugin(category="trimesh", requires=['Rhino'])
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    return K
+
+
+@pluggable(category="trimesh", requires=['Rhino'])
```
```suggestion
@plugin(category="trimesh", requires=['Rhino'])
```

beverlylytle commented on src/compas_rhino/geometry/trimesh/curvature.py:
```diff
@@ -0,0 +1,331 @@
+    return H
+
+
+@pluggable(category="trimesh", requires=['Rhino'])
```
```suggestion
@plugin(category="trimesh", requires=['Rhino'])
```

beverlylytle reviewed (changes requested):
Hi @XingxinHE !  The math and the code look really good!  I hope you are not overwhelmed by the number of comments.  They are mostly about little changes to the documentation.  But there are one or two changes that need to be made before we merge this.  In particular, the change from `pluggable` to `plugin` in the three cases that I commented on.  And I'd like to know your thoughts on why you chose zero as the default value in degenerate situations like when `vert_neighbors_topo` is `None`.  We really appreciate your contribution!

XingxinHE committed:
Revise code in light of feedback.
1. Modify CHANGELOG
2. Modify Subscript i on math formula
3. Add one condition to detect if this is a trimesh
4. Add dihedral angle reference
5. Modify sqrt with determinant
6. Modify the return value of isolated vertex as None
7. Modify some grammars of docs

XingxinHE commented:
Hi @beverlylytle  Dr. Lytle, 

What I saw is not "overwhelm" comments but your patience and useful feedback, I really appreciate that!  OK, below is a summary of the commit.

1. Modify CHANGELOG
2. Modify Subscript i of math formula
3. Add one condition to detect if this is a trimesh
    > I agree with you. How about adding a condition to see if the input is a `triangular mesh`? With this condition, the function itself is more consistent with the tab `trimesh`.
4. Add dihedral angle reference
    > I added it right after the formula of mean curvature which is compact I think.
5. Modify sqrt with determinant
    > Thx for the tips :-)
6. Modify the return value of isolated vertex as None
    > This is a good question! In the beginning, I defined the isolated vertex as zero curvature due to the post-processing after the calculation. For example, by adding a shader to the vertex, the user doesn't need to check if this is `None` or not. and use it as it is.... But I totally agree with you. The `None` curvature makes more sense and it is more reasonable for an isolated vertex,
7. Modify some grammars of docs
    > Thx.....

That's all! My sincere thanks again :-)

beverlylytle reviewed (approved):
Looks great! I'll resolve the conflicts and then merge.  Thanks for your efforts!

beverlylytle committed:
Merge branch 'main' into trimesh_rhino

beverlylytle committed:
Clean up docstrings, fix imports
